// @azure/keyvault-secrets/dist/commonjs/lro/recover/operation.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var operation_exports = {};
__export(operation_exports, {
  RecoverDeletedSecretPollOperation: () => RecoverDeletedSecretPollOperation
});
module.exports = __toCommonJS(operation_exports);
var import_keyVaultSecretPoller = require("../keyVaultSecretPoller.js");
var import_transformations = require("../../transformations.js");
var import_tracing = require("../../tracing.js");
class RecoverDeletedSecretPollOperation extends import_keyVaultSecretPoller.KeyVaultSecretPollOperation {
  constructor(state, client, options = {}) {
    super(state, { cancelMessage: "Canceling the recovery of a deleted secret is not supported." });
    this.state = state;
    this.client = client;
    this.options = options;
  }
  state;
  client;
  options;
  /**
   * The getSecret method returns the specified secret along with its properties.
   * This operation requires the secrets/get permission.
   */
  getSecret(name, options = {}) {
    return import_tracing.tracingClient.withSpan(
      "RecoverDeletedSecretPoller.getSecret",
      options,
      async (updatedOptions) => {
        const response = await this.client.getSecret(
          name,
          options && options.version ? options.version : "",
          updatedOptions
        );
        return (0, import_transformations.getSecretFromSecretBundle)(response);
      }
    );
  }
  /**
   * The recoverDeletedSecret method recovers the specified deleted secret along with its properties.
   * This operation requires the secrets/recover permission.
   */
  recoverDeletedSecret(name, options = {}) {
    return import_tracing.tracingClient.withSpan(
      "RecoverDeletedSecretPoller.recoverDeletedSecret",
      options,
      async (updatedOptions) => {
        const response = await this.client.recoverDeletedSecret(name, updatedOptions);
        return (0, import_transformations.getSecretFromSecretBundle)(response);
      }
    );
  }
  /**
   * Reaches to the service and updates the delete secret's poll operation.
   */
  async update(options = {}) {
    const state = this.state;
    const { name } = state;
    if (options.abortSignal) {
      this.options.abortSignal = options.abortSignal;
    }
    if (!state.isStarted) {
      try {
        state.result = (await this.getSecret(name, this.options)).properties;
        state.isCompleted = true;
      } catch {
      }
      if (!state.isCompleted) {
        state.result = (await this.recoverDeletedSecret(name, this.options)).properties;
        state.isStarted = true;
      }
    }
    if (!state.isCompleted) {
      try {
        state.result = (await this.getSecret(name, this.options)).properties;
        state.isCompleted = true;
      } catch (error) {
        if (error.statusCode === 403) {
          state.isCompleted = true;
        } else if (error.statusCode !== 404) {
          state.error = error;
          state.isCompleted = true;
          throw error;
        }
      }
    }
    return this;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  RecoverDeletedSecretPollOperation
});
