// @azure/keyvault-secrets/dist/commonjs/api/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var api_exports = {};
__export(api_exports, {
  backupSecret: () => import_operations.backupSecret,
  createKeyVault: () => import_keyVaultContext.createKeyVault,
  deleteSecret: () => import_operations.deleteSecret,
  getDeletedSecret: () => import_operations.getDeletedSecret,
  getDeletedSecrets: () => import_operations.getDeletedSecrets,
  getSecret: () => import_operations.getSecret,
  getSecretVersions: () => import_operations.getSecretVersions,
  getSecrets: () => import_operations.getSecrets,
  purgeDeletedSecret: () => import_operations.purgeDeletedSecret,
  recoverDeletedSecret: () => import_operations.recoverDeletedSecret,
  restoreSecret: () => import_operations.restoreSecret,
  setSecret: () => import_operations.setSecret,
  updateSecret: () => import_operations.updateSecret
});
module.exports = __toCommonJS(api_exports);
var import_keyVaultContext = require("./keyVaultContext.js");
var import_operations = require("./operations.js");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  backupSecret,
  createKeyVault,
  deleteSecret,
  getDeletedSecret,
  getDeletedSecrets,
  getSecret,
  getSecretVersions,
  getSecrets,
  purgeDeletedSecret,
  recoverDeletedSecret,
  restoreSecret,
  setSecret,
  updateSecret
});
