// @azure/keyvault-secrets/dist/commonjs/api/operations.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var operations_exports = {};
__export(operations_exports, {
  _backupSecretDeserialize: () => _backupSecretDeserialize,
  _backupSecretSend: () => _backupSecretSend,
  _deleteSecretDeserialize: () => _deleteSecretDeserialize,
  _deleteSecretSend: () => _deleteSecretSend,
  _getDeletedSecretDeserialize: () => _getDeletedSecretDeserialize,
  _getDeletedSecretSend: () => _getDeletedSecretSend,
  _getDeletedSecretsDeserialize: () => _getDeletedSecretsDeserialize,
  _getDeletedSecretsSend: () => _getDeletedSecretsSend,
  _getSecretDeserialize: () => _getSecretDeserialize,
  _getSecretSend: () => _getSecretSend,
  _getSecretVersionsDeserialize: () => _getSecretVersionsDeserialize,
  _getSecretVersionsSend: () => _getSecretVersionsSend,
  _getSecretsDeserialize: () => _getSecretsDeserialize,
  _getSecretsSend: () => _getSecretsSend,
  _purgeDeletedSecretDeserialize: () => _purgeDeletedSecretDeserialize,
  _purgeDeletedSecretSend: () => _purgeDeletedSecretSend,
  _recoverDeletedSecretDeserialize: () => _recoverDeletedSecretDeserialize,
  _recoverDeletedSecretSend: () => _recoverDeletedSecretSend,
  _restoreSecretDeserialize: () => _restoreSecretDeserialize,
  _restoreSecretSend: () => _restoreSecretSend,
  _setSecretDeserialize: () => _setSecretDeserialize,
  _setSecretSend: () => _setSecretSend,
  _updateSecretDeserialize: () => _updateSecretDeserialize,
  _updateSecretSend: () => _updateSecretSend,
  backupSecret: () => backupSecret,
  deleteSecret: () => deleteSecret,
  getDeletedSecret: () => getDeletedSecret,
  getDeletedSecrets: () => getDeletedSecrets,
  getSecret: () => getSecret,
  getSecretVersions: () => getSecretVersions,
  getSecrets: () => getSecrets,
  purgeDeletedSecret: () => purgeDeletedSecret,
  recoverDeletedSecret: () => recoverDeletedSecret,
  restoreSecret: () => restoreSecret,
  setSecret: () => setSecret,
  updateSecret: () => updateSecret
});
module.exports = __toCommonJS(operations_exports);
var import_models = require("../models/models.js");
var import_pagingHelpers = require("../static-helpers/pagingHelpers.js");
var import_urlTemplate = require("../static-helpers/urlTemplate.js");
var import_core_client = require("@azure-rest/core-client");
function _restoreSecretSend(context, parameters, options = { requestOptions: {} }) {
  const path = (0, import_urlTemplate.expandUrlTemplate)(
    "/secrets/restore{?api%2Dversion}",
    {
      "api%2Dversion": context.apiVersion
    },
    {
      allowReserved: options?.requestOptions?.skipUrlEncoding
    }
  );
  return context.path(path).post({
    ...(0, import_core_client.operationOptionsToRequestParameters)(options),
    contentType: "application/json",
    headers: {
      accept: "application/json",
      ...options.requestOptions?.headers
    },
    body: (0, import_models.secretRestoreParametersSerializer)(parameters)
  });
}
async function _restoreSecretDeserialize(result) {
  const expectedStatuses = ["200"];
  if (!expectedStatuses.includes(result.status)) {
    const error = (0, import_core_client.createRestError)(result);
    error.details = (0, import_models.keyVaultErrorDeserializer)(result.body);
    throw error;
  }
  return (0, import_models.secretBundleDeserializer)(result.body);
}
async function restoreSecret(context, parameters, options = { requestOptions: {} }) {
  const result = await _restoreSecretSend(context, parameters, options);
  return _restoreSecretDeserialize(result);
}
function _backupSecretSend(context, secretName, options = { requestOptions: {} }) {
  const path = (0, import_urlTemplate.expandUrlTemplate)(
    "/secrets/{secret-name}/backup{?api%2Dversion}",
    {
      "secret-name": secretName,
      "api%2Dversion": context.apiVersion
    },
    {
      allowReserved: options?.requestOptions?.skipUrlEncoding
    }
  );
  return context.path(path).post({
    ...(0, import_core_client.operationOptionsToRequestParameters)(options),
    headers: {
      accept: "application/json",
      ...options.requestOptions?.headers
    }
  });
}
async function _backupSecretDeserialize(result) {
  const expectedStatuses = ["200"];
  if (!expectedStatuses.includes(result.status)) {
    const error = (0, import_core_client.createRestError)(result);
    error.details = (0, import_models.keyVaultErrorDeserializer)(result.body);
    throw error;
  }
  return (0, import_models.backupSecretResultDeserializer)(result.body);
}
async function backupSecret(context, secretName, options = { requestOptions: {} }) {
  const result = await _backupSecretSend(context, secretName, options);
  return _backupSecretDeserialize(result);
}
function _recoverDeletedSecretSend(context, secretName, options = { requestOptions: {} }) {
  const path = (0, import_urlTemplate.expandUrlTemplate)(
    "/deletedsecrets/{secret-name}/recover{?api%2Dversion}",
    {
      "secret-name": secretName,
      "api%2Dversion": context.apiVersion
    },
    {
      allowReserved: options?.requestOptions?.skipUrlEncoding
    }
  );
  return context.path(path).post({
    ...(0, import_core_client.operationOptionsToRequestParameters)(options),
    headers: {
      accept: "application/json",
      ...options.requestOptions?.headers
    }
  });
}
async function _recoverDeletedSecretDeserialize(result) {
  const expectedStatuses = ["200"];
  if (!expectedStatuses.includes(result.status)) {
    const error = (0, import_core_client.createRestError)(result);
    error.details = (0, import_models.keyVaultErrorDeserializer)(result.body);
    throw error;
  }
  return (0, import_models.secretBundleDeserializer)(result.body);
}
async function recoverDeletedSecret(context, secretName, options = { requestOptions: {} }) {
  const result = await _recoverDeletedSecretSend(context, secretName, options);
  return _recoverDeletedSecretDeserialize(result);
}
function _purgeDeletedSecretSend(context, secretName, options = { requestOptions: {} }) {
  const path = (0, import_urlTemplate.expandUrlTemplate)(
    "/deletedsecrets/{secret-name}{?api%2Dversion}",
    {
      "secret-name": secretName,
      "api%2Dversion": context.apiVersion
    },
    {
      allowReserved: options?.requestOptions?.skipUrlEncoding
    }
  );
  return context.path(path).delete({
    ...(0, import_core_client.operationOptionsToRequestParameters)(options),
    headers: {
      accept: "application/json",
      ...options.requestOptions?.headers
    }
  });
}
async function _purgeDeletedSecretDeserialize(result) {
  const expectedStatuses = ["204"];
  if (!expectedStatuses.includes(result.status)) {
    const error = (0, import_core_client.createRestError)(result);
    error.details = (0, import_models.keyVaultErrorDeserializer)(result.body);
    throw error;
  }
  return;
}
async function purgeDeletedSecret(context, secretName, options = { requestOptions: {} }) {
  const result = await _purgeDeletedSecretSend(context, secretName, options);
  return _purgeDeletedSecretDeserialize(result);
}
function _getDeletedSecretSend(context, secretName, options = { requestOptions: {} }) {
  const path = (0, import_urlTemplate.expandUrlTemplate)(
    "/deletedsecrets/{secret-name}{?api%2Dversion}",
    {
      "secret-name": secretName,
      "api%2Dversion": context.apiVersion
    },
    {
      allowReserved: options?.requestOptions?.skipUrlEncoding
    }
  );
  return context.path(path).get({
    ...(0, import_core_client.operationOptionsToRequestParameters)(options),
    headers: {
      accept: "application/json",
      ...options.requestOptions?.headers
    }
  });
}
async function _getDeletedSecretDeserialize(result) {
  const expectedStatuses = ["200"];
  if (!expectedStatuses.includes(result.status)) {
    const error = (0, import_core_client.createRestError)(result);
    error.details = (0, import_models.keyVaultErrorDeserializer)(result.body);
    throw error;
  }
  return (0, import_models.deletedSecretBundleDeserializer)(result.body);
}
async function getDeletedSecret(context, secretName, options = { requestOptions: {} }) {
  const result = await _getDeletedSecretSend(context, secretName, options);
  return _getDeletedSecretDeserialize(result);
}
function _getDeletedSecretsSend(context, options = { requestOptions: {} }) {
  const path = (0, import_urlTemplate.expandUrlTemplate)(
    "/deletedsecrets{?api%2Dversion,maxresults}",
    {
      "api%2Dversion": context.apiVersion,
      maxresults: options?.maxresults
    },
    {
      allowReserved: options?.requestOptions?.skipUrlEncoding
    }
  );
  return context.path(path).get({
    ...(0, import_core_client.operationOptionsToRequestParameters)(options),
    headers: {
      accept: "application/json",
      ...options.requestOptions?.headers
    }
  });
}
async function _getDeletedSecretsDeserialize(result) {
  const expectedStatuses = ["200"];
  if (!expectedStatuses.includes(result.status)) {
    const error = (0, import_core_client.createRestError)(result);
    error.details = (0, import_models.keyVaultErrorDeserializer)(result.body);
    throw error;
  }
  return (0, import_models._deletedSecretListResultDeserializer)(result.body);
}
function getDeletedSecrets(context, options = { requestOptions: {} }) {
  return (0, import_pagingHelpers.buildPagedAsyncIterator)(
    context,
    () => _getDeletedSecretsSend(context, options),
    _getDeletedSecretsDeserialize,
    ["200"],
    { itemName: "value", nextLinkName: "nextLink" }
  );
}
function _getSecretVersionsSend(context, secretName, options = { requestOptions: {} }) {
  const path = (0, import_urlTemplate.expandUrlTemplate)(
    "/secrets/{secret-name}/versions{?api%2Dversion,maxresults}",
    {
      "secret-name": secretName,
      "api%2Dversion": context.apiVersion,
      maxresults: options?.maxresults
    },
    {
      allowReserved: options?.requestOptions?.skipUrlEncoding
    }
  );
  return context.path(path).get({
    ...(0, import_core_client.operationOptionsToRequestParameters)(options),
    headers: {
      accept: "application/json",
      ...options.requestOptions?.headers
    }
  });
}
async function _getSecretVersionsDeserialize(result) {
  const expectedStatuses = ["200"];
  if (!expectedStatuses.includes(result.status)) {
    const error = (0, import_core_client.createRestError)(result);
    error.details = (0, import_models.keyVaultErrorDeserializer)(result.body);
    throw error;
  }
  return (0, import_models._secretListResultDeserializer)(result.body);
}
function getSecretVersions(context, secretName, options = { requestOptions: {} }) {
  return (0, import_pagingHelpers.buildPagedAsyncIterator)(
    context,
    () => _getSecretVersionsSend(context, secretName, options),
    _getSecretVersionsDeserialize,
    ["200"],
    { itemName: "value", nextLinkName: "nextLink" }
  );
}
function _getSecretsSend(context, options = { requestOptions: {} }) {
  const path = (0, import_urlTemplate.expandUrlTemplate)(
    "/secrets{?api%2Dversion,maxresults}",
    {
      "api%2Dversion": context.apiVersion,
      maxresults: options?.maxresults
    },
    {
      allowReserved: options?.requestOptions?.skipUrlEncoding
    }
  );
  return context.path(path).get({
    ...(0, import_core_client.operationOptionsToRequestParameters)(options),
    headers: {
      accept: "application/json",
      ...options.requestOptions?.headers
    }
  });
}
async function _getSecretsDeserialize(result) {
  const expectedStatuses = ["200"];
  if (!expectedStatuses.includes(result.status)) {
    const error = (0, import_core_client.createRestError)(result);
    error.details = (0, import_models.keyVaultErrorDeserializer)(result.body);
    throw error;
  }
  return (0, import_models._secretListResultDeserializer)(result.body);
}
function getSecrets(context, options = { requestOptions: {} }) {
  return (0, import_pagingHelpers.buildPagedAsyncIterator)(
    context,
    () => _getSecretsSend(context, options),
    _getSecretsDeserialize,
    ["200"],
    { itemName: "value", nextLinkName: "nextLink" }
  );
}
function _getSecretSend(context, secretName, secretVersion, options = { requestOptions: {} }) {
  const path = (0, import_urlTemplate.expandUrlTemplate)(
    "/secrets/{secret-name}/{secret-version}{?api%2Dversion,outContentType}",
    {
      "secret-name": secretName,
      "secret-version": secretVersion,
      "api%2Dversion": context.apiVersion,
      outContentType: options?.outContentType
    },
    {
      allowReserved: options?.requestOptions?.skipUrlEncoding
    }
  );
  return context.path(path).get({
    ...(0, import_core_client.operationOptionsToRequestParameters)(options),
    headers: {
      accept: "application/json",
      ...options.requestOptions?.headers
    }
  });
}
async function _getSecretDeserialize(result) {
  const expectedStatuses = ["200"];
  if (!expectedStatuses.includes(result.status)) {
    const error = (0, import_core_client.createRestError)(result);
    error.details = (0, import_models.keyVaultErrorDeserializer)(result.body);
    throw error;
  }
  return (0, import_models.secretBundleDeserializer)(result.body);
}
async function getSecret(context, secretName, secretVersion, options = { requestOptions: {} }) {
  const result = await _getSecretSend(context, secretName, secretVersion, options);
  return _getSecretDeserialize(result);
}
function _updateSecretSend(context, secretName, secretVersion, parameters, options = { requestOptions: {} }) {
  const path = (0, import_urlTemplate.expandUrlTemplate)(
    "/secrets/{secret-name}/{secret-version}{?api%2Dversion}",
    {
      "secret-name": secretName,
      "secret-version": secretVersion,
      "api%2Dversion": context.apiVersion
    },
    {
      allowReserved: options?.requestOptions?.skipUrlEncoding
    }
  );
  return context.path(path).patch({
    ...(0, import_core_client.operationOptionsToRequestParameters)(options),
    contentType: "application/json",
    headers: {
      accept: "application/json",
      ...options.requestOptions?.headers
    },
    body: (0, import_models.secretUpdateParametersSerializer)(parameters)
  });
}
async function _updateSecretDeserialize(result) {
  const expectedStatuses = ["200"];
  if (!expectedStatuses.includes(result.status)) {
    const error = (0, import_core_client.createRestError)(result);
    error.details = (0, import_models.keyVaultErrorDeserializer)(result.body);
    throw error;
  }
  return (0, import_models.secretBundleDeserializer)(result.body);
}
async function updateSecret(context, secretName, secretVersion, parameters, options = { requestOptions: {} }) {
  const result = await _updateSecretSend(context, secretName, secretVersion, parameters, options);
  return _updateSecretDeserialize(result);
}
function _deleteSecretSend(context, secretName, options = { requestOptions: {} }) {
  const path = (0, import_urlTemplate.expandUrlTemplate)(
    "/secrets/{secret-name}{?api%2Dversion}",
    {
      "secret-name": secretName,
      "api%2Dversion": context.apiVersion
    },
    {
      allowReserved: options?.requestOptions?.skipUrlEncoding
    }
  );
  return context.path(path).delete({
    ...(0, import_core_client.operationOptionsToRequestParameters)(options),
    headers: {
      accept: "application/json",
      ...options.requestOptions?.headers
    }
  });
}
async function _deleteSecretDeserialize(result) {
  const expectedStatuses = ["200"];
  if (!expectedStatuses.includes(result.status)) {
    const error = (0, import_core_client.createRestError)(result);
    error.details = (0, import_models.keyVaultErrorDeserializer)(result.body);
    throw error;
  }
  return (0, import_models.deletedSecretBundleDeserializer)(result.body);
}
async function deleteSecret(context, secretName, options = { requestOptions: {} }) {
  const result = await _deleteSecretSend(context, secretName, options);
  return _deleteSecretDeserialize(result);
}
function _setSecretSend(context, secretName, parameters, options = { requestOptions: {} }) {
  const path = (0, import_urlTemplate.expandUrlTemplate)(
    "/secrets/{secret-name}{?api%2Dversion}",
    {
      "secret-name": secretName,
      "api%2Dversion": context.apiVersion
    },
    {
      allowReserved: options?.requestOptions?.skipUrlEncoding
    }
  );
  return context.path(path).put({
    ...(0, import_core_client.operationOptionsToRequestParameters)(options),
    contentType: "application/json",
    headers: {
      accept: "application/json",
      ...options.requestOptions?.headers
    },
    body: (0, import_models.secretSetParametersSerializer)(parameters)
  });
}
async function _setSecretDeserialize(result) {
  const expectedStatuses = ["200"];
  if (!expectedStatuses.includes(result.status)) {
    const error = (0, import_core_client.createRestError)(result);
    error.details = (0, import_models.keyVaultErrorDeserializer)(result.body);
    throw error;
  }
  return (0, import_models.secretBundleDeserializer)(result.body);
}
async function setSecret(context, secretName, parameters, options = { requestOptions: {} }) {
  const result = await _setSecretSend(context, secretName, parameters, options);
  return _setSecretDeserialize(result);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  _backupSecretDeserialize,
  _backupSecretSend,
  _deleteSecretDeserialize,
  _deleteSecretSend,
  _getDeletedSecretDeserialize,
  _getDeletedSecretSend,
  _getDeletedSecretsDeserialize,
  _getDeletedSecretsSend,
  _getSecretDeserialize,
  _getSecretSend,
  _getSecretVersionsDeserialize,
  _getSecretVersionsSend,
  _getSecretsDeserialize,
  _getSecretsSend,
  _purgeDeletedSecretDeserialize,
  _purgeDeletedSecretSend,
  _recoverDeletedSecretDeserialize,
  _recoverDeletedSecretSend,
  _restoreSecretDeserialize,
  _restoreSecretSend,
  _setSecretDeserialize,
  _setSecretSend,
  _updateSecretDeserialize,
  _updateSecretSend,
  backupSecret,
  deleteSecret,
  getDeletedSecret,
  getDeletedSecrets,
  getSecret,
  getSecretVersions,
  getSecrets,
  purgeDeletedSecret,
  recoverDeletedSecret,
  restoreSecret,
  setSecret,
  updateSecret
});
