// @softeria/ms-365-mcp-server/dist/logger.js
import winston from "winston";
import path from "path";
import { fileURLToPath } from "url";
import fs from "fs";
import os from "os";
const __dirname = path.dirname(fileURLToPath(import.meta.url));
const logsDir = process.env.MS365_MCP_LOG_DIR || path.join(os.homedir(), ".ms-365-mcp-server", "logs");
if (!fs.existsSync(logsDir)) {
  fs.mkdirSync(logsDir, { recursive: true, mode: 448 });
}
const logger = winston.createLogger({
  level: process.env.LOG_LEVEL || "info",
  format: winston.format.combine(
    winston.format.timestamp({
      format: "YYYY-MM-DD HH:mm:ss"
    }),
    winston.format.printf(({ level, message, timestamp }) => {
      return `${timestamp} ${level.toUpperCase()}: ${message}`;
    })
  ),
  transports: [
    new winston.transports.File({
      filename: path.join(logsDir, "error.log"),
      level: "error"
    }),
    new winston.transports.File({
      filename: path.join(logsDir, "mcp-server.log")
    })
  ]
});
const enableConsoleLogging = () => {
  logger.add(
    new winston.transports.Console({
      format: winston.format.combine(winston.format.colorize(), winston.format.simple()),
      silent: process.env.SILENT === "true" || process.env.SILENT === "1"
    })
  );
};
var logger_default = logger;
export {
  logger_default as default,
  enableConsoleLogging
};
