// @azure/keyvault-secrets/dist/commonjs/keyVaultClient.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var keyVaultClient_exports = {};
__export(keyVaultClient_exports, {
  KeyVaultClient: () => KeyVaultClient
});
module.exports = __toCommonJS(keyVaultClient_exports);
var import_api = require("./api/index.js");
var import_operations = require("./api/operations.js");
class KeyVaultClient {
  _client;
  /** The pipeline used by this client to make requests */
  pipeline;
  /** The key vault client performs cryptographic key operations and vault operations against the Key Vault service. */
  constructor(endpointParam, credential, options = {}) {
    const prefixFromOptions = options?.userAgentOptions?.userAgentPrefix;
    const userAgentPrefix = prefixFromOptions ? `${prefixFromOptions} azsdk-js-client` : `azsdk-js-client`;
    this._client = (0, import_api.createKeyVault)(endpointParam, credential, {
      ...options,
      userAgentOptions: { userAgentPrefix }
    });
    this.pipeline = this._client.pipeline;
  }
  /** Restores a backed up secret, and all its versions, to a vault. This operation requires the secrets/restore permission. */
  restoreSecret(parameters, options = { requestOptions: {} }) {
    return (0, import_operations.restoreSecret)(this._client, parameters, options);
  }
  /** Requests that a backup of the specified secret be downloaded to the client. All versions of the secret will be downloaded. This operation requires the secrets/backup permission. */
  backupSecret(secretName, options = { requestOptions: {} }) {
    return (0, import_operations.backupSecret)(this._client, secretName, options);
  }
  /** Recovers the deleted secret in the specified vault. This operation can only be performed on a soft-delete enabled vault. This operation requires the secrets/recover permission. */
  recoverDeletedSecret(secretName, options = { requestOptions: {} }) {
    return (0, import_operations.recoverDeletedSecret)(this._client, secretName, options);
  }
  /** The purge deleted secret operation removes the secret permanently, without the possibility of recovery. This operation can only be enabled on a soft-delete enabled vault. This operation requires the secrets/purge permission. */
  purgeDeletedSecret(secretName, options = { requestOptions: {} }) {
    return (0, import_operations.purgeDeletedSecret)(this._client, secretName, options);
  }
  /** The Get Deleted Secret operation returns the specified deleted secret along with its attributes. This operation requires the secrets/get permission. */
  getDeletedSecret(secretName, options = { requestOptions: {} }) {
    return (0, import_operations.getDeletedSecret)(this._client, secretName, options);
  }
  /** The Get Deleted Secrets operation returns the secrets that have been deleted for a vault enabled for soft-delete. This operation requires the secrets/list permission. */
  getDeletedSecrets(options = { requestOptions: {} }) {
    return (0, import_operations.getDeletedSecrets)(this._client, options);
  }
  /** The full secret identifier and attributes are provided in the response. No values are returned for the secrets. This operations requires the secrets/list permission. */
  getSecretVersions(secretName, options = { requestOptions: {} }) {
    return (0, import_operations.getSecretVersions)(this._client, secretName, options);
  }
  /** The Get Secrets operation is applicable to the entire vault. However, only the base secret identifier and its attributes are provided in the response. Individual secret versions are not listed in the response. This operation requires the secrets/list permission. */
  getSecrets(options = { requestOptions: {} }) {
    return (0, import_operations.getSecrets)(this._client, options);
  }
  /** The GET operation is applicable to any secret stored in Azure Key Vault. This operation requires the secrets/get permission. */
  getSecret(secretName, secretVersion, options = { requestOptions: {} }) {
    return (0, import_operations.getSecret)(this._client, secretName, secretVersion, options);
  }
  /** The UPDATE operation changes specified attributes of an existing stored secret. Attributes that are not specified in the request are left unchanged. The value of a secret itself cannot be changed. This operation requires the secrets/set permission. */
  updateSecret(secretName, secretVersion, parameters, options = { requestOptions: {} }) {
    return (0, import_operations.updateSecret)(this._client, secretName, secretVersion, parameters, options);
  }
  /** The DELETE operation applies to any secret stored in Azure Key Vault. DELETE cannot be applied to an individual version of a secret. This operation requires the secrets/delete permission. */
  deleteSecret(secretName, options = { requestOptions: {} }) {
    return (0, import_operations.deleteSecret)(this._client, secretName, options);
  }
  /** The SET operation adds a secret to the Azure Key Vault. If the named secret already exists, Azure Key Vault creates a new version of that secret. This operation requires the secrets/set permission. */
  setSecret(secretName, parameters, options = { requestOptions: {} }) {
    return (0, import_operations.setSecret)(this._client, secretName, parameters, options);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  KeyVaultClient
});
