// @azure/keyvault-secrets/dist/commonjs/transformations.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var transformations_exports = {};
__export(transformations_exports, {
  getSecretFromSecretBundle: () => getSecretFromSecretBundle,
  mapPagedAsyncIterable: () => mapPagedAsyncIterable
});
module.exports = __toCommonJS(transformations_exports);
var import_identifier = require("./identifier.js");
function getSecretFromSecretBundle(bundle) {
  const secretBundle = bundle;
  const deletedSecretBundle = bundle;
  const parsedId = (0, import_identifier.parseKeyVaultSecretIdentifier)(secretBundle.id);
  const attributes = secretBundle.attributes;
  delete secretBundle.attributes;
  const resultObject = {
    value: secretBundle.value,
    name: parsedId.name,
    properties: {
      expiresOn: attributes?.expires,
      createdOn: attributes?.created,
      updatedOn: attributes?.updated,
      enabled: attributes?.enabled,
      notBefore: attributes?.notBefore,
      recoverableDays: attributes?.recoverableDays,
      recoveryLevel: attributes?.recoveryLevel,
      id: secretBundle.id,
      contentType: secretBundle.contentType,
      tags: secretBundle.tags,
      managed: secretBundle.managed,
      previousVersion: secretBundle.previousVersion,
      vaultUrl: parsedId.vaultUrl,
      version: parsedId.version,
      name: parsedId.name,
      certificateKeyId: secretBundle.kid
    }
  };
  if (deletedSecretBundle.recoveryId) {
    resultObject.properties.recoveryId = deletedSecretBundle.recoveryId;
    resultObject.properties.scheduledPurgeDate = deletedSecretBundle.scheduledPurgeDate;
    resultObject.properties.deletedOn = deletedSecretBundle.deletedDate;
    resultObject.recoveryId = deletedSecretBundle.recoveryId;
    resultObject.scheduledPurgeDate = deletedSecretBundle.scheduledPurgeDate;
    resultObject.deletedOn = deletedSecretBundle.deletedDate;
  }
  if (attributes) {
    if (attributes.vaultUrl) {
      delete resultObject.properties.vaultUrl;
    }
    if (attributes.expires) {
      delete resultObject.properties.expires;
    }
    if (attributes.created) {
      delete resultObject.properties.created;
    }
    if (attributes.updated) {
      delete resultObject.properties.updated;
    }
  }
  return resultObject;
}
function mapPagedAsyncIterable(operation, operationOptions, mapper) {
  let iter = void 0;
  return {
    async next() {
      iter ??= operation({ ...operationOptions, maxresults: void 0 });
      const result = await iter.next();
      return {
        ...result,
        value: result.value && mapper(result.value)
      };
    },
    [Symbol.asyncIterator]() {
      return this;
    },
    async *byPage(settings) {
      const iteratorByPage = operation({
        ...operationOptions,
        maxresults: settings?.maxPageSize
      }).byPage(settings);
      for await (const page of iteratorByPage) {
        yield page.map(mapper);
      }
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getSecretFromSecretBundle,
  mapPagedAsyncIterable
});
